pandas/tests/extension/decimal/array.py

```diff
@@ -78,6 +78,15 @@ def take(self, indexer, allow_fill=True, fill_value=None):
         indexer = np.asarray(indexer)
         mask = indexer == -1
 
+        # take on empty array
+        if not len(self):
+            # only valid if result is an all-missing array
+            if mask.all():
```

Comment:
writing take is a fair amount of effort on the extension writer. meaning additional complexity and have to handle edge cases. take is a pure indexing operation and so ideally should be implemetned as high up in the stack as possible (meaning in the base class for EA) if the EA writer wants to override then ok, but it should just work generally. All of the machinery is already there. The point of the EA is to make it easy, having to handle all of these edge cases is not easy.

Reply:
> All of the machinery is already there.

Assuming you have an ndarray. Do we want to provide a default implementation that does an `astype(object)` to convert to an ndarray of scalars, `take` on that, and then construct a new EA?

I'm going to make a PR documenting which methods use `.astype(object)`, so that the EA author should consider overriding them if performance is a concern.

Reply:
> Assuming you have an ndarray. Do we want to provide a default implementation that does an astype(object) to convert to an ndarray of scalars, take on that, and then construct a new EA?

It will almost never be what you want I think. It could still serve as an example implementation of course, like we now have one in the docstring.